app/src/main/java/ch/epfl/polybazaar/database/Model.java

```diff
@@ -0,0 +1,52 @@
+package ch.epfl.polybazaar.database;
+
+import com.google.android.gms.tasks.Task;
+import com.google.android.gms.tasks.Tasks;
+
+import ch.epfl.polybazaar.database.datastore.DataStore;
+import ch.epfl.polybazaar.database.datastore.DataStoreFactory;
+
+public abstract class Model {
+    /**
+     * Returns the name of the collection corresponding to the model
+     * @return model name
+     */
+    public abstract String collectionName();
+
+    /**
+     * Returns the id of the model instance
+     * @return model id
+     */
+    public abstract String getId();
+
+    /**
+     * Sets the id of the model instance
+     * @param id
+     */
+    public abstract void setId(String id);
+
+    /**
+     * Saves the model on the database
+     * @return successful task if the database was reachable
+     */
+    public final Task<Void> save() {
+        DataStore db = DataStoreFactory.getDependency();
+        if (getId() == null) {
+            return db.addData(collectionName(), this)
+                    .onSuccessTask(s -> {
+                        setId(s);
+                        return Tasks.forResult(null);
+                    });
+        } else {
+            return db.setData(collectionName(), getId(), this);
+        }
+    }
+
```

Comment:
Is there a way to have `fetch` as an abstract method knowing that the return type depends of the class?

Reply:
Oups, `fetch` is a static method.

Reply:
It is unfortunately not very convenient to do in Java because we have to pass the class as an argument, but it would indeed be better if we did not have to add this method manually to every model.